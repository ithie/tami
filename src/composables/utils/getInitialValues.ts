import { ref, type Ref } from 'vue'

export default <T = Ref<number>>(
  [energy, happieness, hunger]: [number, number, number],
  noRef = false
): { energy: T; happieness: T; hunger: T } => ({
  energy: (noRef ? energy : ref(energy)) as T,
  happieness: (noRef ? happieness : ref(happieness)) as T,
  hunger: (noRef ? hunger : ref(hunger)) as T
})
