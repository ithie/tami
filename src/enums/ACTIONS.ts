export enum ACTIONS {
  FEED = 'FEED',
  PLAY = 'PLAY',
  SLEEP = 'SLEEP',
  NOTHING = 'NOTHING'
}
