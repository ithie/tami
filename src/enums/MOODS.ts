export enum MOODS {
  NEUTRAL = 'NEUTRAL',
  HUNGRY = 'HUNGRY',
  VERY_HUNGRY = 'VERY_HUNGRY',
  HAPPY = 'HAPPY',
  SUPER_HAPPY = 'SUPER_HAPPY',
  TIRED = 'TIRED'
}
