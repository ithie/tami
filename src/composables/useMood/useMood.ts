import { ref } from 'vue'
import { ACTIONS } from '../../enums/ACTIONS'
import restrictRanges from '../utils/restrictRanges'

const hunger = ref(5)
const happieness = ref(5)
const energy = ref(5)
const age = ref(1)

const restrictToMaximum = () => {
  restrictRanges(hunger)
  restrictRanges(happieness)
  restrictRanges(energy)
}

const ACTION_HANDLES = {
  [ACTIONS.FEED]: () => {
    hunger.value -= 2
    happieness.value += 1
    energy.value += 2
  },
  [ACTIONS.PLAY]: () => {
    happieness.value = happieness.value + 2
    energy.value -= 1
    hunger.value += 1
  },
  [ACTIONS.SLEEP]: () => {
    hunger.value += 3
    energy.value += 2
  }
}

export default (vals?: { hunger?: number; happieness?: number; energy?: number; age?: number }) => {
  if (vals) {
    hunger.value = vals.hunger || 5
    happieness.value = vals.happieness || 5
    energy.value = vals.energy || 5
    age.value = vals.age || 1
  }

  return {
    doAction: (action: ACTIONS) => {
      age.value++
      hunger.value += 1
      happieness.value -= 1
      energy.value -= 1

      if (action !== ACTIONS.NOTHING) {
        ACTION_HANDLES[action]()
      }

      restrictToMaximum()
    },
    age,
    hunger,
    happieness,
    energy
  }
}
