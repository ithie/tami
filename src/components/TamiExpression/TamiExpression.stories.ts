import type { Meta, StoryObj } from '@storybook/vue3'
import TamiExpression from './TamiExpression.vue'
import { MOODS } from '../../enums/MOODS'
import useMood from '../../composables/useMood/useMood'

const moodValues = {
  [MOODS.NEUTRAL]: {
    happieness: 5,
    hunger: 5,
    energy: 5
  },
  [MOODS.HUNGRY]: {
    happieness: 5,
    hunger: 7,
    energy: 5
  },
  [MOODS.VERY_HUNGRY]: {
    happieness: 5,
    hunger: 9,
    energy: 5
  },
  [MOODS.HAPPY]: {
    happieness: 7,
    hunger: 5,
    energy: 5
  },
  [MOODS.SUPER_HAPPY]: {
    happieness: 9,
    hunger: 5,
    energy: 5
  },
  [MOODS.TIRED]: {
    happieness: 5,
    hunger: 5,
    energy: 0
  }
}

const meta = {
  title: 'Components/TamiExpression',
  component: TamiExpression,
  render: (args) => ({
    setup() {
      console.log(args.state)
      if (!args.state || !moodValues[args.state]) {
        useMood({ happieness: 5, hunger: 5, energy: 5 })
      } else {
        useMood(moodValues[args.state])
      }
    },
    components: { TamiExpression },
    template: '<TamiExpression />'
  }),
  parameters: {},
  args: {},
  // This component will have an automatically generated docsPage entry: https://storybook.js.org/docs/writing-docs/autodocs
  tags: ['autodocs']
} satisfies Meta<{
  state?: MOODS
}>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {}

export const ExpressionNeutral: Story = {
  args: {
    state: MOODS.NEUTRAL
  }
}

export const ExpressionHungry: Story = {
  args: {
    state: MOODS.HUNGRY
  }
}

export const ExpressionVeryHungry: Story = {
  args: {
    state: MOODS.VERY_HUNGRY
  }
}

export const ExpressionHappy: Story = {
  args: {
    state: MOODS.HAPPY
  }
}

export const ExpressionSuperHappy: Story = {
  args: {
    state: MOODS.SUPER_HAPPY
  }
}

export const ExpressionTired: Story = {
  args: {
    state: MOODS.TIRED
  }
}
