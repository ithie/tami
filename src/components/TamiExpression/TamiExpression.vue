<template>
  <div class="tami-border">
    <pre>{{ tami }}</pre>
  </div>
</template>

<script lang="ts" setup>
import useRenderMood from '../../composables/useRenderMood/useRenderMood'
import useMood from '../../composables/useMood/useMood'

const { hunger, happieness, energy } = useMood()

const tami = useRenderMood({ hunger, happieness, energy })
</script>

<style lang="scss">
.tami-border {
  text-align: center;
  background-color: gainsboro;
  border: 1px solid black;
}
</style>
