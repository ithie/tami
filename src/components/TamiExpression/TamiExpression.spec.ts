import { describe, test, expect } from 'vitest'
import { flushPromises, mount } from '@vue/test-utils'
import TamiExpression from './TamiExpression.vue'
import useMood from '../../composables/useMood/useMood'
import getInitialValues from '../../composables/utils/getInitialValues'
import { MOODS } from '../../enums/MOODS'

const assertions: Array<[MOODS, { hunger: number; happieness: number; energy: number }]> = [
  [MOODS.TIRED, getInitialValues<number>([0, 5, 5], true)],
  [MOODS.VERY_HUNGRY, getInitialValues<number>([5, 9, 5], true)],
  [MOODS.TIRED, getInitialValues<number>([5, 5, 3], true)],
  [MOODS.HUNGRY, getInitialValues<number>([5, 5, 7], true)],
  [MOODS.SUPER_HAPPY, getInitialValues<number>([5, 9, 5], true)],
  [MOODS.HAPPY, getInitialValues<number>([5, 7, 5], true)],
  [MOODS.HAPPY, getInitialValues<number>([5, 5, 5], true)]
]

describe('components/TamiExpression', () => {
  assertions.forEach(([assertion, given]) => {
    test(`should render mood: ${assertion}`, async () => {
      const wrapper = mount(TamiExpression)

      useMood(given)
      await flushPromises()
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})
