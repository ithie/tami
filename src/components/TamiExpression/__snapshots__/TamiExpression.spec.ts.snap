// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`components/TamiExpression > should render mood: HAPPY 1`] = `
<div
  class="tami-border"
>
  <pre>
    (* *)
  </pre>
</div>
`;

exports[`components/TamiExpression > should render mood: HAPPY 2`] = `
<div
  class="tami-border"
>
  <pre>
    (o o)
  </pre>
</div>
`;

exports[`components/TamiExpression > should render mood: HUNGRY 1`] = `
<div
  class="tami-border"
>
  <pre>
    (- - )
  </pre>
</div>
`;

exports[`components/TamiExpression > should render mood: SUPER_HAPPY 1`] = `
<div
  class="tami-border"
>
  <pre>
    (^ ^)
  </pre>
</div>
`;

exports[`components/TamiExpression > should render mood: TIRED 1`] = `
<div
  class="tami-border"
>
  <pre>
    (o o)
  </pre>
</div>
`;

exports[`components/TamiExpression > should render mood: TIRED 2`] = `
<div
  class="tami-border"
>
  <pre>
    (u u)
  </pre>
</div>
`;

exports[`components/TamiExpression > should render mood: VERY_HUNGRY 1`] = `
<div
  class="tami-border"
>
  <pre>
    (^ ^)
  </pre>
</div>
`;
